File: style-labs/kick-forward-style/TermFrequency.js
// Kick-Forward Style

const fs = require("fs");
const { argv } = require("process");

function readFile(pathToFile, func) {
	func(fs.readFileSync(pathToFile, "utf8"));
}

function cleanText(text, func) {
	func(text.replace(/[\W_]+/g, " "));
}

function removeSingleCharWords(text, func) {
	func(text.replace(/\b\w{1}\b/g, ""));
}

function toLowerText(text, func) {
	func(text.toLowerCase());
}

function createWordsArray(text, func) {
	func(text.split(" "));
}

function removeStopWords(words, stopWords, func) {
	func(words.filter((e) => !stopWords.includes(e)));
}

function wordsFrequency(words, func) {
	const wordsFreq = new Map();
	words.forEach((word) => {
		wordsFreq.set(word, (wordsFreq.get(word) ?? 0) + 1);
	});
	func(wordsFreq);
}

function sortMapByFreq(wordsFreqMap, func) {
	const sortedEntries = [...wordsFreqMap.entries()].sort((a, b) => b[1] - a[1]);
	func(new Map(sortedEntries));
}

function limit(maxNumber, words, func) {
	func([...words].slice(0, maxNumber));
}

function wordsFrequencyToString(wordsFreq, func) {
	func(
		wordsFreq.reduce((accumulator, [word, frequency]) => {
			return accumulator + `${word}  -  ${frequency}\n`;
		}, "")
	);
}

readFile(argv[2], (text) =>
	cleanText(text, (cleaned) =>
		removeSingleCharWords(cleaned, (noSingleChar) =>
			toLowerText(noSingleChar, (lowerText) =>
				createWordsArray(lowerText, (words) =>
					readFile("stop_words.txt", (stopWordsText) =>
						cleanText(stopWordsText, (cleanStopWordsText) =>
							removeSingleCharWords(cleanStopWordsText, (noSingleCharStop) =>
								toLowerText(noSingleCharStop, (lowerStopText) =>
									createWordsArray(lowerStopText, (stopWords) =>
										removeStopWords(words, stopWords, (wordsWithoutStopWords) =>
											wordsFrequency(wordsWithoutStopWords, (wordsFreq) =>
												sortMapByFreq(wordsFreq, (sortedWordsFreq) =>
													limit(argv[3], sortedWordsFreq, (limitRes) =>
														wordsFrequencyToString(limitRes, console.log)
													)
												)
											)
										)
									)
								)
							)
						)
					)
				)
			)
		)
	)
);

File: TermFrequency.java
// Kick-Forward Style

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class TermFrequency {

    private static void readFile(final String pathToFile, Consumer<String> f) {
        try {
            f.accept(Files.readString(Paths.get(pathToFile), StandardCharsets.UTF_8));
        } catch (final IOException ex) {
            return;
        }
    }

    private static void cleanText(final String text, Consumer<String> f) {
        f.accept(text.replaceAll("[\\W_]+", " "));
    }

    private static void removeSingleCharWords(final String text, Consumer<String> f) {
        f.accept(text.replaceAll("\\b\\w{1}\\b", ""));
    }

    private static void toLowerCase(final String text, Consumer<String> f) {
        f.accept(text.toLowerCase());
    }

    private static void splitTextInWords(final String text, Consumer<String[]> f) {
        f.accept(text.split(" "));
    }

    private static void removeStopWords(final String[] words, final String[] stopWords, Consumer<String[]> f) {
        final Set<String> stopWordsSet = Arrays.stream(stopWords).collect(Collectors.toSet());
        stopWordsSet.add("");
        final List<String> filteredWords = Arrays.stream(words).filter(word -> !stopWordsSet.contains(word))
                .collect(Collectors.toList());
        f.accept(filteredWords.toArray(new String[0]));
    }

    private static void countWordFrequency(final String[] words, Consumer<HashMap<String, Integer>> f) {
        final HashMap<String, Integer> wordsFreqMap = new HashMap<>();

        for (String word : words) {
            if (!wordsFreqMap.containsKey(word))
                wordsFreqMap.put(word, 1);
            else
                wordsFreqMap.replace(word, wordsFreqMap.get(word) + 1);
        }

        f.accept(wordsFreqMap);
    }

    private static <R, S> void sortHashMap(final HashMap<R, S> map, Consumer<LinkedHashMap<R, S>> f) {
        f.accept(map.entrySet()
                .stream()
                .sorted(Map.Entry.comparingByValue(Collections.reverseOrder()))
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        Map.Entry::getValue,
                        (e1, e2) -> e1,
                        LinkedHashMap::new)));
    }

    private static <R, S> void sliceHashMap(final LinkedHashMap<R, S> map, final Integer maxSize,
            Consumer<LinkedHashMap<R, S>> f) {
        f.accept(map.entrySet()
                .stream()
                .limit(maxSize)
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        Map.Entry::getValue,
                        (e1, e2) -> e1,
                        LinkedHashMap::new)));
    }

    private static <R, S> void hashMapToString(final LinkedHashMap<R, S> map, Consumer<String> f) {
        final StringBuilder sb = new StringBuilder();
        map.forEach((word, frequency) -> sb.append(word + "  -  " + frequency + "\n"));
        f.accept(sb.toString());
    }

    public static void main(String[] args) {
        //TODO: remove stopwords
        readFile(args[0], text -> cleanText(text, cleanedText -> removeSingleCharWords(cleanedText,
                noSingleCharText -> toLowerCase(noSingleCharText, lowerCaseText -> splitTextInWords(lowerCaseText,
                        words -> countWordFrequency(words,
                                wordsFreq -> sortHashMap(wordsFreq,
                                        sortedWordsFreq -> sliceHashMap(sortedWordsFreq, Integer.parseInt(args[1]),
                                                limitWordsFreq -> hashMapToString(limitWordsFreq,
                                                        stringResult -> System.out.println(stringResult))))))))));
    }
}